### ISSUE TITLE
[FEAT][ELEMENTWISE] independent ops with custom signatures (11 ops)

### ISSUE THREAD

**Issue by lcy-seso** [tile-ai/TileOPs#439]
## 1. Feature Scope

11 ops with non-uniform signatures. Each inherits `Kernel` ABC directly — no shared template base class. Shared coding pattern: global→register, `T.Parallel`, all in `elementwise.py`.

| Op | Extra Params | Input Signature |
|----|-------------|-----------------|
| leaky_relu | negative_slope (scalar) | `(x,) → y` |
| elu | alpha (scalar) | `(x,) → y` |
| hardtanh | min_val, max_val (scalars) | `(x,) → y` |
| softplus | beta, threshold (scalars) | `(x,) → y` |
| prelu | weight (tensor, per-channel) | `(x, weight) → y` |
| where | — | `(cond, x, y) → out` |
| clamp | min, max (scalars, optional) | `(x,) → y` |
| masked_fill | value (scalar) | `(x, mask) → y` |
| nan_to_num | nan, posinf, neginf (scalars) | `(x,) → y` |
| alibi | — | Position encoding specific |
| sinusoidal | — | Position encoding specific |

**Excluded**: relu/gelu/silu/... (template ops, #437 #438).

### Files

| File | Action |
|------|--------|
| `tileops/kernels/elementwise.py` | Add 11 kernel classes |
| `tileops/ops/elementwise.py` | Add 11 op classes |
| `tests/ops/test_activation.py` | Append: leaky_relu, elu, hardtanh, softplus, prelu |
| `tests/ops/test_special_elementwise.py` | Append: where, clamp, masked_fill, nan_to_num, alibi, sinusoidal |

### Kernel Pattern (example: leaky_relu)

Each op builds its own kernel function with op-specific params baked in as compile-time constants:

```python
def _make_leaky_relu_kernel(N, dtype, negative_slope, threads=256, npt=8):
    block_size = threads * npt
    @T.prim_func
    def main(x: T.Tensor((N,), dtype), y: T.Tensor((N,), dtype)):
        with T.Kernel(T.ceildiv(N, block_size), threads=threads) as bx:
            for i, j in T.Parallel(threads, npt):
                idx = (bx * threads + i) * npt + j
                y[idx] = T.if_then_else(x[idx] > 0, x[idx], negative_slope * x[idx])
    return main

class LeakyReluKernel(Kernel):
    supported_archs = [80, 86, 89, 90]

    def __init__(self, N_total, dtype, negative_slope=0.01, config=None, tune=False):
        super().__init__()
        self.N_total, self.dtype = N_total, dtype
        self.negative_slope = negative_slope
        cfg = self.default_config
        self.kernel = _make_leaky_relu_kernel(N_total, self.dtype_str, negative_slope, cfg["threads"], cfg["num_per_thread"])
        self.init_config(config, tune)

    @property
    def default_config(self):
        npt = 4 if self.dtype == torch.float32 else 8
        return {"threads": 256, "num_per_thread": npt}

    def forward(self, x):
        return self.kernel(x)
```

### Op-Specific Kernel Logic

| Op | Core `op_func` expression | Notes |
|----|--------------------------|-------|
| leaky_relu | `T.if_then_else(x > 0, x, negative_slope * x)` | |
| elu | `T.if_then_else(x > 0, x, alpha * (T.exp(x) - 1))` | |
| hardtanh | `T.min(T.max(x, min_val), max_val)` | |
| softplus | `T.if_then_else(x * beta > threshold, x, T.log(1 + T.exp(x * beta)) / beta)` | |
| prelu | `T.if_then_else(x > 0, x, weight[channel] * x)` | weight is per-channel tensor; needs channel index calc |
| where | `T.if_then_else(cond[idx], x[idx], y[idx])` | 3 input tensors |
| clamp | `T.min(T.max(x, min_val), max_val)` | min/max optional (None = unbounded) |
| masked_fill | `T.if_then_else(mask[idx], value, x[idx])` | |
| nan_to_num | nested `if_then_else`: replace NaN→nan, +Inf→posinf, -Inf→neginf | |
| alibi | Position-dependent linear bias computation | Research ALiBi paper for exact formula |
| sinusoidal | `sin(pos / 10000^(2i/d))` / `cos(pos / 10000^(2i/d))` | Alternating sin/cos by dimension |

### Op Layer

Each independent op has its own Op class inheriting `Op` ABC, with custom `__init__` accepting op-specific params and custom `forward` handling input reshape.

### L1 — Correctness (11 cases, fp16, 1M)

| Op | PyTorch Reference | Input Notes |
|----|-------------------|-------------|
| leaky_relu | `F.leaky_relu(x, 0.01)` | Default negative_slope |
| elu | `F.elu(x, 1.0)` | Default alpha |
| hardtanh | `F.hardtanh(x, -1, 1)` | Default bounds |
| softplus | `F.softplus(x, 1, 20)` | Default beta/threshold |
| prelu | `F.prelu(x, weight)` | Random weight tensor |
| where | `torch.where(cond, x, y)` | Random bool cond |
| clamp | `torch.clamp(x, -0.5, 0.5)` | Both bounds |
| masked_fill | `x.masked_fill(mask, -1e9)` | Random bool mask |
| nan_to_num | `torch.nan_to_num(x, 0, 1e4, -1e4)` | Input with NaN/Inf injected |
| alibi | Manual reference computation | |
| sinusoidal | Manual reference computation | |

### L2 — Dtype x Size (4 cases, representative: clamp)

| Dtype | Size |
|-------|------|
| fp32, bf16 | 1M |
| fp16 | 4K, 16M |

### L4 — Edge Cases (8 cases, fp32, 4K)

| Op | Input | Expected |
|----|-------|----------|
| clamp | min > max | Defined behavior per PyTorch (min wins) |
| clamp | min=None, max=0.5 | Upper bound only |
| clamp | min=-0.5, max=None | Lower bound only |
| masked_fill | all True mask | All values replaced |
| masked_fill | all False mask | Input unchanged |
| where | all True cond | Output = x |
| where | all False cond | Output = y |
| nan_to_num | `[NaN, Inf, -Inf, 1.0]` | `[0, 1e4, -1e4, 1.0]` |

## 4. Risks

| Risk | Mitigation |
|------|------------|
| prelu per-channel weight indexing adds complexity | Need channel dim calculation in kernel; test with multi-channel input |
| alibi/sinusoidal formulas need external reference | Verify against paper / existing implementations before coding |
| No template base → more code per op (~30 lines each vs 3-5) | Accept this; the ops genuinely have different signatures |
| softplus threshold branch may cause numerical issues | Edge case test with values near threshold boundary |

## 5. Acceptance Criteria

- [ ] 11 kernel + 11 op classes implemented (each inheriting `Kernel`/`Op` ABC)
- [ ] `__all__` updated in both `elementwise.py`
- [ ] L1: 11 ops pass (fp16, 1M)
- [ ] L4: 8 edge cases pass (fp32, 4K)
- [ ] `__init__.py` not modified
